department-of-veterans-affairs/va.gov-cms
# VACMS-14816 Add Vet Center outdated content email

## Description

Relates to #14816

### Primary changes

1. Refactor Vet Center to use content templates
2. Refactor to allow different subjects by product.
3. Alter email text and subject
4. Refactor layout to make accessible at zoom other than 100%. Makes it mobile friendly.
5. Refactor to make it easier to add other products.
6. Add the ability to test with a subset of editors.
7. Add extensive documentation for future products and other teams.
8. **Bugfix:** Digging for outdated content gave up if the first one it tried had none.  This made it possible that some editors are not receiving notifications if their first qualifying section was up-to-date or was a section that was outside the product. 
9. **Bugfix:** VAMC Banner image not pathed in all environments.  Emails sent could not load the banner.  It had no domain. 
![image](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/5752113/25135ead-cfe7-4cb0-b058-bfbeb6aa02eb)


## Testing done
## QA steps
Be aware:  If testing this locally, you will be able to see the email in full html.  Unfortunately Tugboat's mail catcher is text only.
As an admin
1. [Go here](https://pr15232-h8mncqapl0pvwsxsw7tdzvjxerkfdf98.ci.cms.va.gov/user/4480/workbench_access) and assign this user to the following sections
   - VBA facilities (a section that should get no emails)
   -  VA Bedford health care ( a section that should receive VAMC email)
   - Colorado Springs Vet Center (a vet center with outdated content) 
Save the form.
2. Go to https://pr15232-h8mncqapl0pvwsxsw7tdzvjxerkfdf98.ci.cms.va.gov/user/4660/edit and block them.  (they are blocked on prod, but not on all lower environments.
3. Navigate to the [queue list](https://pr15232-h8mncqapl0pvwsxsw7tdzvjxerkfdf98.ci.cms.va.gov/admin/config/system/queues) and take note of the queued items for both "VAMC Oudated Content" and "Vet Center Outdated Content"  Most likely will be 0 for each unless someone else has been in testing as well.
4. In the next steps you will be using the following user ids.
   - 4480 ( G. Ferra, The editor you edited, someone who is now a member of both vba, VAMC and Vet center - checking for perfect storm)
   - 3244 (L. Li, an editor with no VAMC or vet center section)
   - 1812 (C. Lyons, a VAMC VISN 2 level editor)
   - 4323 (M. Barrier, a single VAMC system editor)
   - 4660 (B. Martin, a blocked VAMC editor)
   - 3259 (C. Jones, a vet center editor)
5. Go to [this code's tugboat](https://tugboat.vfs.va.gov/6501add32d3036648d6201e5) and click on the link for the php terminal
6. Test VAMC email, in the terminal run `drush php-eval "print_r(\Drupal::service('va_gov_notifications.outdated_content')->queueOutdatedContentNotifications('vamc', 'vamc_outdated_content', ['4480', '3244', '1812', '4323', '4660', '3259']));"`
   - [x] validate that response is:
```
Array
(
    [0] => Array
        (
            [editor] => Gerald.Ferra@va.gov
            [section] => VA Bedford health care
        )

    [1] => Array
        (
            [editor] => Cullen.lyons@va.gov
            [section] => VA Albany health care
        )

    [2] => Array
        (
            [editor] => Marlin.Barrier@va.gov
            [section] => VA Philadelphia health care
        )
)
```
These are the people who should be getting VAMC email.

   7. Test Vet Center email, in the terminal run `drush php-eval "print_r(\Drupal::service('va_gov_notifications.outdated_content')->queueOutdatedContentNotifications('vet_center', 'vet_center_outdated_content', ['4480', '3244', '1812', '4323', '4660', '3259']));"`
   - [x] validate that response is :
```
Array
(
    [0] => Array
        (
            [editor] => Gerald.Ferra@va.gov
            [section] => Colorado Springs Vet Center
        )

    [1] => Array
        (
            [editor] => clifford.jones1@va.gov
            [section] => South Jersey Vet Center
        )

)
```
These are the people who should be getting Vet Center email.

8. Visit the [queue list here](https://pr15232-h8mncqapl0pvwsxsw7tdzvjxerkfdf98.ci.cms.va.gov/admin/config/system/queues):
   - [x] Validate there are 3 more items in the VAMC outdated content queue
   - [x] Validate there are 2 more items in the Vet Center outdated content queue.
9. Process the queues by going to the terminal and running `drush cron`.
10. Visit the [captured mail section in tugboat or mailhog locally](https://tugboat.vfs.va.gov/6501add32d3036648d6201e5).
   - [x] Validate there are 2 vet center emails
   - [x] Validate there are 3 VAMC emails.
11. Open a Vet Center email
   - [x] validate the subject is "[ACTION REQUIRED] Vet Center Website Content Review".
   - [x] validate the body reads correctly
   - [x] Validate the banner image is visible(locally only)  on tugboat, view the source, scroll down to the html section and validate that the image src has a domain and full path (you should be able copy and paste the url into an new tab and see the banner image.
11. Open a VAMC email
   - [x] validate the subject is "[ACTION REQUIRED] Review Content Report".
   - [x] validate the body reads correctly
   - [x] Validate the banner image is visible.
   - [x] Validate that the links work
   - [x] validate that it scales


## Post deploy steps
1. SSH to prod 
- [ ] 2. Queue the sample VAMC emails to select editors with this command 
```
drush php-eval "print_r(\Drupal::service('va_gov_notifications.outdated_content')->queueOutdatedContentNotifications('vamc', 'vamc_outdated_content', ['1342', '4433', '4650', '4017', '4634', '4375', '4305', '4073', '4216', '4197', '4222', '1215']));"
```
- [ ] 3. Queue the sample VetCenter emails to select editors with this command 
```
drush php-eval "print_r(\Drupal::service('va_gov_notifications.outdated_content')->queueOutdatedContentNotifications('vet_center', 'vet_center_outdated_content', ['1342', '4433', '4650', '4017', '4634', '4375', '4305', '4073', '4216', '4197', '4222', '1215']));"
```
- [ ]  4. Do the sending (process the queues) with `drush cron`
- [ ] 5. Notify in [this slack thread](https://dsva.slack.com/archives/C0FQSS30V/p1695129544542599) "Both a VAMC and Vet Center email have been sent."  Please see them in your VA.gov email and report back here ASAP if you see any issues."
- [ ] 6. If the emails check out correctly, we are done.  If not, we'll need to followup with a fix before Oct 1
